flax/linen/__init__.py

```diff
@@ -151,3 +151,17 @@
     while_loop as while_loop,
 )
 # pylint: enable=g-multiple-import
+
+
+# aliasing the deprecated MultiHeadDotProductAttention to MultiHeadAttention and raising a DeprecationWarning
+def __getattr__(attr):
+  if attr == 'MultiHeadDotProductAttention':
+    import warnings
+    warnings.warn(
+      'flax.linen.MultiHeadDotProductAttention is deprecated. Use flax.linen.MultiHeadAttention instead.',
+      DeprecationWarning,
+      stacklevel=2,
+    )
+    del warnings
+    return MultiHeadAttention
+  raise AttributeError(f"module '{__name__}' has no attribute '{attr}'")
```

Comment:
To improve IDE usability its nice to use `typing.TYPE_CHECKING` to statically express what the API will look like on type checking, and the implement the actual behavior for runtime on the else branch e.g:

```python
import typing as _typing

if _typing.TYPE_CHECKING:
  from .attention import MultiHeadAttention as MultiHeadDotProductAttention
else:
  def __getattr__(attr: str):
    ...
```